core/histogram.cpp

```diff
@@ -6,71 +6,129 @@
 #include "scipp/common/numeric.h"
 #include "scipp/core/dataset.h"
 #include "scipp/core/except.h"
-#include "scipp/core/transform.h"
+#include "scipp/core/transform_subspan.h"
 
 #include "dataset_operations_common.h"
 
 namespace scipp::core {
 
-// For now this implementation is only for the simplest case of 2 dims (inner
-// stands for sparse)
+static constexpr auto make_histogram = [](auto &data, const auto &events,
+                                          const auto &edges) {
+  if (scipp::numeric::is_linspace(edges)) {
+    // Special implementation for linear bins. Gives a 1x to 20x speedup
+    // for few and many events per histogram, respectively.
+    const auto [offset, nbin, scale] = linear_edge_params(edges);
+    for (const auto &e : events) {
+      const double bin = (e - offset) * scale;
+      if (bin >= 0.0 && bin < nbin)
+        ++data.value[static_cast<scipp::index>(bin)];
+    }
+  } else {
+    expect::histogram::sorted_edges(edges);
+    for (const auto &e : events) {
+      auto it = std::upper_bound(edges.begin(), edges.end(), e);
+      if (it != edges.end() && it != edges.begin())
+        ++data.value[--it - edges.begin()];
+    }
+  }
+  std::copy(data.value.begin(), data.value.end(), data.variance.begin());
+};
+
+static constexpr auto make_histogram_from_weighted =
+    [](auto &data, const auto &events, const auto &weights, const auto &edges) {
+      if (scipp::numeric::is_linspace(edges)) {
+        const auto [offset, nbin, scale] = linear_edge_params(edges);
+        for (scipp::index i = 0; i < scipp::size(events); ++i) {
+          const auto x = events[i];
+          const double bin = (x - offset) * scale;
+          if (bin >= 0.0 && bin < nbin) {
+            const auto b = static_cast<scipp::index>(bin);
+            const auto w = weights.values[i];
+            const auto e = weights.variances[i];
+            data.value[b] += w;
+            data.variance[b] += e;
+          }
+        }
+      } else {
+        expect::histogram::sorted_edges(edges);
+        for (scipp::index i = 0; i < scipp::size(events); ++i) {
+          const auto x = events[i];
+          auto it = std::upper_bound(edges.begin(), edges.end(), x);
+          if (it != edges.end() && it != edges.begin()) {
+            const auto b = --it - edges.begin();
+            const auto w = weights.values[i];
+            const auto e = weights.variances[i];
+            data.value[b] += w;
+            data.variance[b] += e;
+          }
+        }
+      }
+    };
```

Comment:
I don't feel this is providing a major simplification, and it would require additional benchmarking to ensure that the `optional` is optimized out, since this is in performance-critical code.